File: GroupAnagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, ArrayList> map = new HashMap<>();
        for(int i=0; i<strs.length; i++){
            char[] ch = strs[i].toCharArray();
            Arrays.sort(ch);
            String key = String.valueOf(ch);
            if(!map.containsKey(key)){
                map.put(key, new ArrayList<String>());
            }
            map.get(key).add(strs[i]);
        }
        return new ArrayList(map.values());
    }
}

File: GroupAnagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
 // TC: O(N * N) SC: O(N * K)
var groupAnagrams = function(strs) {
    const anagramMap = new Map();
    strs.forEach((el) => {
        const count = new Array(26).fill(0);
        for(let i=0; i<el.length; i++) {
            count[el.charCodeAt(i) - 'a'.charCodeAt(0)]++;
        }
        const anagram = count.join('-');
        if(anagramMap.has(anagram)) {
            anagramMap.get(anagram).push(el);
        }
        else {
            anagramMap.set(anagram, [el]);
        }
    })
    return [...anagramMap.values()];


/**
 * @param {string[]} strs
 * @return {string[][]}
 */
 // TC: O(N * KlogK) SC: O(N * K)
var groupAnagrams = function(strs) {
    const anagramMap = new Map();
    strs.forEach((el) => {
        const anagram = el.split('').sort().join('');
        if(anagramMap.has(anagram)) {
            anagramMap.get(anagram).push(el);
        }
        else {
            anagramMap.set(anagram, [el]);
        }
    })
    return [...anagramMap.values()];
};


